Added DeepImageFeaturizer and tests.

MrBago opened the pull request:
We're anticipating switching the image schema to spark.image.ImageSchema (https://issues.apache.org/jira/browse/SPARK-21866).

Until that schema is merged we'll be depending on microsoft/spark-image. There is a small difference in the schema that'll need to be fixed when the change happens.

MrBago committed:
Added DeepImageFeaturizer and tests.

MrBago committed:
Small clarity change.

MrBago committed:
Update featurizer to use updated ImageUtils.

MrBago committed:
Merge branch 'image-utils' into deep-image-featurizer-scala

MrBago committed:
Test bigger & samller image with DeepImageFeaturizer.

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeatureizer.scala:
Nit: Rename this file from "DeepImageFeatureizer.scala" -> "DeepImageFeaturizer.scala"

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeatureizer.scala:
Should we perform schema validation here? E.g. check the datatypes of `inputCol` & `outputCol` to make sure they're both strings? Here's an example of how it's done in Spark: https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala#L151.

smurching commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
I'm a little confused by this test; how should DeepImageFeaturizer behave when its output column is already present in the DataFrame passed to transform? AFAICT the transformer currently overwrites existing columns if needed. See the docs for Dataset.withColumn: https://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.Dataset

smurching commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
My recommendation would be to fail if an output column already exists, which would match the behavior in existing Spark transformers (`StringIndexer`, `Bucketizer`, `HashingTF`, etc)

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeatureizer.scala:
Follow-up: If we do add schema validation, we should unit test it.

smurching reviewed (commented):
A few small comments -- besides nits, I had a few Qs that we can discuss in person if you'd like (to spare you from typing out responses).

MrBago committed:
Merge branch 'image-utils' into deep-image-featurizer-scala

MrBago committed:
PR feedback.

MrBago committed:
Better schema validation and errors.

MrBago commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
Sorry this was leftover from when I was trying to do something very unnecessary. I've added better errors and a better test.

smurching commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Nit: `validatSchema` -> `validateSchema`. Otherwise this LGTM!

MrBago committed:
Fix typo.

sueann commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
just to confirm, this runs on the Dataset sent into transform right

sueann commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
why is this called `Debug...`? xD @thunterdb

sueann commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
do you need `"Test named image featurizer runs on our image samples"` given this test?

sueann reviewed (commented):
just a few questions for you @MrBago and @thunterdb. would be good for tim to check the tensorframes part but i'm okay with this as long as the tests run fine.

MrBago commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Yes @thunterdb, why is it named Debug... :).

MrBago commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Yep. It's also used in `transformSchema` which, my understanding is, is used by Pipelines to ensure that the pipeline is valid (maybe during a call to pipeline.transform?).

MrBago committed:
Merge branch 'image-utils' into deep-image-featurizer-scala

Minor changes to fix tests with updated resource directories.

MrBago commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
The "out image samples" test is more of an integration test, it ensures that our featurizer works with a data frame created by "readImages" and makes sure that in can handle images of different sizes and shapes. I can take it out if you think it's too much.

sueann commented on src/test/scala/com/databricks/sparkdl/DeepImageFeaturizerSuite.scala:
ah ok. maybe rename it to "Test named image featurizer runs on image dataframes" ?

sueann reviewed (approved):
lgtm. just one small comment. thanks!

MrBago committed:
Merge branch 'image-utils' into deep-image-featurizer-scala

MrBago committed:
Rename test.

thunterdb commented on python/sparkdl/transformers/keras_applications.py:
We should merge that upstream, separately. This is a comment for the future, no need to act on it.

thunterdb commented on python/tests/transformers/named_image_test.py:
We should address this issue in tensorframes too, this is annoying.

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
do we need this to be public?

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Yes, this is because the pipeline API tries to be more lazy than what Spark allows, especially when combining estimators into a pipeline.

Nit by the way: `: Unit = {` As written, it will not be compatible with Scala 3.

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
indeed, it can infer it from the graph.

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Well, it is called debug, because it was written with idea that it would be soon replaced by `OptimizedRowOps`, but got optimized enough in the process...

You are not supposed to access this internal function, but at the same time it seems useful to expose it.

thunterdb commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
`.filter` instead

thunterdb commented on .travis.yml:
this is not used in this PR. I really want to refactor `run-tests.sh` into a python script with a real interface eventually.

thunterdb commented:
@MrBago some tiny comments. Otherwise I am looking forward to it.

MrBago commented on python/sparkdl/transformers/keras_applications.py:
Sorry that was my mistake, it only showed up on this PR because i failed to push the image-utils branch last night :(.

MrBago commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
this is just a placeholder for the param. I've fixed it #2 .

MrBago commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Is there any benefit to me trying to put the right thing here?

MrBago commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Scala 3 requires all methods to have return types?

MrBago committed:
Tim's PR comments.

MrBago commented on src/main/scala/com/databricks/sparkdl/DeepImageFeaturizer.scala:
Tim what should out be here? Is it ok if I leave it as empty?

MrBago commented:
@thunterdb I think I've addressed your comments, take a look when you get a chance.

btw the plan is to merge this "down", meaning we should start merging with #2 .

thunterdb commented:
@MrBago LGTM

MrBago committed:
InceptionV3 & model download logic. (#2)

* Added InceptionV3 model to DeepImageFeaturizer & download logic for
fetching models.

* PR Feedback.

* Proper variabls names.

* Clean up not needed case in getFromWeb catch block.

* Split out modelName Param tests & related code cleanup.

* Moved FetchModel tests into a new Suite.

* Do model fetch testing not using "prod" model cache directory.
